자판기 기본 동작 구현

GangWoon opened the pull request:
코드를 작성하면서 같은 물건이 다른 날짜에 추가될 수 있는데, 그부분을 구현하지 못했습니다.
Beverage 배열을 객체로 따로 만들었는데 제대로 된 선택이였는지 고민했습니다.
음료수의 상, 하위 클래스를 설정할때 공통되는 특징을 찾는데 노력했습니다. 
이름을 작성할때 사전을 믿는게 아닌 제 의도를 녹여내려구 노력했습니다.

GangWoon committed:
DateGenerator 클래스 제거

기본 클래스에서 추가하는 방법을 사용하기 위해서 제거했습니다. 코드를 작성할 때 다른 사람이 볼 수 있다고 생각하고 신중하게 작성하겠습니다.

GangWoon committed:
Date 확장

customDateFormatToString이라는 속성을 만들어서 DateGenerator가 처리했던 일을 수행하도록 했습니다.

GangWoon committed:
Beverage 클래스 생성

ElementsOfBeverage 클래스 이름을 Beverage로 변경했습니다. 그에따른 다른 클래스에서 사용했던 ElementsOfBeverage을 Beverage 바꿔주는 작업을 했습니다.

GangWoon committed:
ViewController 클래스 수정

테스트 코드를 추가해서 viewcontroller에 추가할 필요가 없다고 생각했습니다.

GangWoon committed:
BananaMilk 클래스 삭제

더 이상 바나나 우유가 필요 없어져서 삭제했습니다. 대신 바나나 우유를 대채할 클래스를 생성했습니다.

GangWoon committed:
Beverage 클래스 생성

구글 스위프트 가이드라인을 보면서 public을 붙이기 시작했습니다. init()이 길어지면서 어떻게 해야 잘 보일까 고민하다가 그냥 파라미터가 4개 이상되면 한칸씩 띄어쓰기로 결정했습니다.
Elements of Beverage 이름대신 Beverage로 변경했습니다.
dateOfManufacture대신 expirationDate로 유통기한을 표시하기 위해서 변경했습니다.
온도, 칼로리를 추가했습니다. 추가된 속성을 사용하는 computed properties를 추가했습니다.
hashable을 추가한 이유는 자판기에서 Beverage를 키값으로 사용하기 위해서 채택했습니다.

GangWoon committed:
OilBeverage 클래스 추가

기름 음료라는 의미로 지었습니다.
좀 웃긴 이름이긴 하지만 우유가 되기전 상위 클래스를 작성해야했고, 마땅한 이름이 떠오르지 않아서 이렇게 작성했습니다.

GangWoon committed:
CaffineBeverage 클래스 작성

음료에서 카페인을 추가적인 속성을 갖기때문에 CaffineBeverage로 작성했습니다.

GangWoon committed:
CarbonicAcidBeverage 클래스 생성

concentration은 농도라는 의미를 채택해서 사용했습니다. 저도 잘 이해가 되진 않지만 다른 단어들은 더 이상해서 고민하다가 작성했습니다.

GangWoon committed:
Coffee 클래스 생성

CaffineBeverage를 상속받는 커피 클래스를 생성했습니다.
커피 품종을 속성으로 갖습니다.

GangWoon committed:
Cola 클래스 생성

CarbonicAcidBeverage를 상속받는 콜라를 생성했습니다.
추가적으로 설탕을 상속받아서 일반 콜라인지 다이어트 콜라인지 구별할 수 있습니다.

GangWoon committed:
컨밴션 수정

public을 붙였습니다

GangWoon committed:
Milk 클래스 생성

oilbeverage를 상속받는 milk 클래스를 생성했습니다.
**맛 우유를 생각해서 맛 속성을 같습니다.

GangWoon committed:
파일 삭제

GangWoon committed:
Products 구조체 생성

beverage 배열을 사용하기 때문에 따로 products라는 구조체를 생성했습니다. 여기에서 product라는 말은 없지만 표현의 한계점을 느끼며 products는 배열 product는 아이템 하나로 처리했습니다.
map함수를 구현하고 싶어서 시도를 했었고(throws랑 rethrows를 아직도 잘 모르겠습니다.), 동작도 했었지만 map이 아닌 compactMap이 필요했습니다. compactMap은 구현하지 못해서 돌고돌아서 forEach를 만들고 상위 객체에서 저장하는 방식으로 구현했습니다.

GangWoon committed:
VendingMachine 클래스 추가

bag라는 가방을 사용해서 구매한 물건들을 저장했습니다. 산 물건들을 배열로 리턴하는 함수를 만들라고 하셨는데 제 생각으론 getter 밖에 생각해내지 못했습니다.
balance는 원래 메소드로 구현하라고 하셨지만, 제 생각에는 computed property를 사용해도 괜찮을 거 같아서 구현했습니다.(건방지게 보일 수 있을 거 같아서.... 정말로 정중하게 표현하고 싶은데 제가 언어를 잘 못써서 죄송합니다...)
products에서 map, compactMap을 잘 구현했으면 map 메소드를 따로 만들 필요가 없었지만, 제가 실패해서 자판기가 사용하는 map을 만들었습니다.
만들면서 고민했던 것은 구매를 할때 열거형으로 받아서 처리할까 생각했지만, 음료수의 종류는 엄청 많기 때문에 따로 열거형을 받도록 하지 않았습니다.
리턴 타입이 [Beverage]가 많아서 따로 객체를 생성할까 생각했지만, 처리를 하는게 아닌 리턴타입이 [Beverage]라서 객체를 만들지 않았습니다 typealias 사용할까 생각했었지만 비슷할 거 같아서 추가하지 않았습니다.

GangWoon committed:
UnitTest 클래스 수정

자판기에서 사용한 메소드들을 확인했습니다.
온도의 기준은 0도보다 크면 따듯한 음료이고, 자판기 코인과 물건의 가격을 비교했습니다.

GangWoon committed:
파일 추가로인한 커밋

GangWoon commented:
func map(_ transform: (Beverage) throws -> Beverage) rethrows -> [Beverage] {
        var result: [Beverage] = []
        for item in products.keys {
            result.append(try transform(item))
        }
        return result
    }

map을 이렇게 구현했었는데 products.map { $0.isHot ? $0 : nil } 처럼 사용하면 nil값이 들어와서 이 이상은 이해를 못해서 구현하지 못했었습니다. 
throws를 하고 rethrows를 하는 이유는 밖에서의 문제이기 때문에 다시 던진다고 이해했습니다.

godrm commented on VendingMachineApp/VendingMachineApp/Models/Beverage.swift:
```diff
@@ -0,0 +1,83 @@
+    private var temperature: Double
+    private var calorie: Double
+    public var isHot: Bool {
+        temperature > 0 ? true : false
```
0도만 넘으면 뜨거운 건가요? 0도면 얼음일꺼 같은데... 
이렇게 비교하고 판단하는 로직은 기준 값이 중요합니다. 
기준 값을 하드코딩 하기 보다는 매개변수로 전달하면 조금 더 유연한 구조가 됩니다. 
따라서 computed property 보다는 메소드가 좋겠죠.

GangWoon replied:
조심하겠습니다!

godrm commented on VendingMachineApp/VendingMachineApp/Models/Beverage.swift:
```diff
@@ -0,0 +1,83 @@
+        temperature > 0 ? true : false
+    }
+    public var isLowerCalorie: Bool {
+        calorie > 200 ? true : false
```
여기도 isHot과 마찬가지네요.

godrm commented on VendingMachineApp/VendingMachineApp/Models/Beverage.swift:
```diff
@@ -0,0 +1,83 @@
+        String(price) + "원, "
+    }
+    public var descriptionName: String {
+        name + ", "
```
description들 따로 뜯어놓은 이유가 뭐죠?

GangWoon replied:
커스텀 스트링을 작성할때 더 유연하게 하려고 그랬습니다. 
사용하지 않아서 삭제했습니다.

godrm commented on VendingMachineApp/VendingMachineApp/Models/Beverage.swift:
```diff
@@ -0,0 +1,83 @@
+    }
+    
+    public func apply(featureOfBeverage: [String]) {
+        featureOfBeverage.forEach { additionalElements.append(", " + $0) }
```
이 함수 의도를 잘 모르겠습니다. 음료수의 feacture 라는게 있나요?

GangWoon replied:
처음에 코드를 작성할 때 추가적인 내용들을 표시하기 위해서 작성했는데, 사용하지 않아서 삭제했습니다.

godrm commented on VendingMachineApp/VendingMachineApp/Models/Beverage.swift:
```diff
@@ -0,0 +1,83 @@
+    }
+    
+    public func balance(payment: Int) -> Int {
+        payment - price
```
이 메소드는 어색합니다. 내부 속성을 바꾸는게 아니라 외부 다른 속성을 바꾸는 것 같은데요?
위에 있는 valid와 함께 금액, 잔액을 처리하는 Money 객체를 설계해보세요.
Int 대신에 돈에 대한 연산, 비교, 타입의 역할을 할 수 있을 것 같네요.

godrm commented on VendingMachineApp/VendingMachineApp/Models/Cola.swift:
```diff
@@ -8,15 +8,35 @@
-    private var pondusHydrogenii: Double
+    private var sugar: Bool
+    public var category: String {
+        sugar ? "일반" : "다이어트"
```
이렇게 비교하는 것을 로직으로 작성할 때는 기준값을 잘 정리해보세요. 
Bool 타입 보다는 양을 표현하는 타입으로 저장하고, 기준값과 비교해서 판단하는 게 좋습니다.
판단하는 결과도 하드코딩하지말고 타입으로 선언해보세요.

godrm commented on VendingMachineApp/VendingMachineApp/Models/OilBeverage.swift:
```diff
@@ -0,0 +1,39 @@
+    
+    private var fat: Bool
+    public var category: String {
+        fat ? "지방" : "저지방"
```
기름 Oil과 fat 지방은 다른 의미가 아닐까요? OilBeverage 라는 표현을 쓰던가요?
모델링을 할 때는 그 분야 도메인에서 사용하는 표현을 쓰는게 좋습니다.

GangWoon replied:
한글로도 뭐라고 표현해야할지 모르겠습니다....

godrm commented on VendingMachineApp/VendingMachineApp/Models/VendingMachine.swift:
```diff
@@ -0,0 +1,66 @@
+        var result: [Beverage] = []
+        products.forEach{ beverage, value in
+            guard let beverage = handler(beverage) else { return }
+            result.append(beverage)
```
이 정도는 forEach를 클로저로 만들기 보다는 filter를 클로저로 만드는 게 어땠을까요?

godrm commented on VendingMachineApp/VendingMachineApp/Models/VendingMachine.swift:
```diff
@@ -0,0 +1,66 @@
+    public mutating func buy(product: Beverage) {
+        guard let value = products[product] else { return }
+        bag.append(product)
+        products.update(value: value - 1, forkey: product)
```
어차피 products 에서 꺼내오는 거라면 한꺼번에 꺼내오고, 갯수도 하나 줄이고 해도 되지 않을까요? 여기서 두 번씩 호출할 필요없지 않을까요?

godrm reviewed (commented):
수동적인 데이터 구조보다 능동적인 객체로 만들려고 좀 더 노력해보세요.
클로저를 만드는 것보다 우선 필요한 메소드를 만드는 것도 시도해봐야 합니다. 
다음 피드백을 확인해보세요.

GangWoon committed:
Date++ 파일 삭제

사용하지 않아서 삭제했습니다.

GangWoon committed:
Money 구조체 생성

자판기에서 돈 관련해서 처리할 객체 money를 생성했습니다.
product의 속성중에 price라는 속성을 private에서 public으로 변경했습니다.
제 생각의 한계여서 선택했습니다.

GangWoon committed:
Beverage 클래스 수정

price의 접근 제한을 변경했습니다.
더 이상 사용하지 않는 속성들을 제거했습니다.
beverage에서 비교를 하는게 아닌 beverage를 갖는 객체에서 비교를 하도록 변경했습니다.

GangWoon committed:
Cola 클래스 수정

콜라를 생성해서 사용하는 객체에서 비교 값을 지정할 수 있도록 변경했습니다.
조금 더 명확하게 판단하기 위해서 설탕의 양을 받도록 수정했습니다.

GangWoon committed:
Products 구조체 수정

compactMap하고 결과만 똑같게 구현했습니다.

GangWoon committed:
VendingMachine  구조체 수정

돈과 관련된 작업은 money 객체를 생성해서 처리하도록 변경했습니다.

GangWoon committed:
VendingMachine 구조체 수정

누락된 코인 추가 메소드를 구현했습니다.

GangWoon committed:
Unit Test 수정

전체적인 구조가 바뀌면서 테스트 코드를 수정했습니다.
파일이 추가됨으로써 변경점도 같이 추가했습니다.

GangWoon commented:
하드코딩을 줄이도록 노력하겠습니다.
판단하는 기준은 그 객체를 사용하는 곳에서 정할 수 있도록 작성하겠습니다!!

godrm commented:
Double , Int 타입보다 더 구체적인 범위, 종류를 결정할 수 있는 타입을 지정하는 것도 방법이 됩니다. 참고하세요.
